Cellular API adapt to device driver model

bjarki-andreasen opened the pull request:
The implementation of the cellular API does not follow the device driver model. This PR fixes that :)

bjarki-andreasen committed:
modem: cellular: Restructure cellular API to fit device driver model

The current cellular API header is not written to conform with
the device driver model. This commit fixes that.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
modem: modem_cellular: Update driver to implement cellular API

Move the implementations of the cellular API in modem_cellular.c
to an cellular_driver_structure, and implement this API structure
withing the device drivers.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

bjarki-andreasen committed:
modem: modem_cellular: Rework cellular_get_signal implementation

The modem_cellular.c driver now uses a common script to get
signal strenght, which is run on demand. This is more efficient
than polling it, and simpler since every modem doesn't have to
define their own variant of the script.

Additionally, the CSQ handler now stores the "raw" rssi value
returned from AT+CSQ to be parsed by the cellular_modem_get_signal
implementation.

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

emillindq commented on drivers/modem/modem_cellular.c:
```diff
@@ -1263,6 +1262,103 @@ static void modem_cellular_cmux_handler(struct modem_cmux *cmux, enum modem_cmux
 	}
 }
 
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(get_signal_csq_chat_script_cmds,
+			      MODEM_CHAT_SCRIPT_CMD_RESP("AT+CSQ", csq_match),
+			      MODEM_CHAT_SCRIPT_CMD_RESP("", ok_match));
+
+MODEM_CHAT_SCRIPT_DEFINE(get_signal_csq_chat_script, get_signal_csq_chat_script_cmds,
+			 abort_matches, modem_cellular_chat_callback_handler, 2);
```
I like this! Resolves the warning :partying_face:

bjarki-andreasen committed:
modem: modem_cellular.c: Remove non standard QCCID handler

The QCCID handler is Quectel specific and as such may not be part
of the modem cellular driver which only supports commands defined
in 3GPP TS 27.007

Signed-off-by: Bjarki Arge Andreasen <bjarki@arge-andreasen.me>

emillindq commented on drivers/modem/modem_cellular.c:
```diff
@@ -350,6 +349,7 @@ static void modem_cellular_chat_on_csq(struct modem_chat *chat, char **argv, uin
 	rssi = atoi(argv[1]);
 
 	if (rssi == 99) {
+		data->rssi = INT16_MAX;
```
Maybe INT16_MIN is more clear as a "reserved number" to signify the RSSI is non-existent?

bjarki-trackunit replied:
I changed the +CSQ handler to simply store the raw falue (0-31 and 99) to be parsed just before returning it to the user with the new function `modem_cellular_csq_parse_rssi`, this way we don't need to choose a reserved value, we can just use 99 like the CSQ does :)

emillindq commented on include/zephyr/drivers/cellular.h:
```diff
@@ -75,6 +76,33 @@ enum cellular_modem_info_type {
+					   const enum cellular_modem_info_type type,
+					   char *info, size_t size);
+
+/** RTC driver API */
+__subsystem struct cellular_driver_api {
+	cellular_api_configure_networks configure_networks;
+	cellular_api_get_supported_networks get_supported_networks;
+	cellular_api_get_signal get_signal;
+	cellular_api_get_modem_info get_modem_info;
+};
```
RTC?

bjarki-trackunit replied:
Changed to Cellular

emillindq commented on drivers/modem/modem_cellular.c:
```diff
@@ -1263,6 +1254,100 @@ static void modem_cellular_cmux_handler(struct modem_cmux *cmux, enum modem_cmux
+	}
+
+	/* Run chat script */
+	switch (type) {
+	case CELLULAR_SIGNAL_RSSI:
+		ret = modem_chat_run_script(&data->chat, &get_signal_csq_chat_script);
+		break;
+
+	case CELLULAR_SIGNAL_RSRP:
+	case CELLULAR_SIGNAL_RSRQ:
+		/* TODO: Run CESQ script */
+		ret = -ENOTSUP;
+		break;
+	}
+
+	/* Verify chat script ran successfully */
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Validate and set received value */
+	switch (type) {
+	case CELLULAR_SIGNAL_RSSI:
+		if (data->rssi == INT16_MAX) {
+			ret = -ENODATA;
+			break;
+		}
+
+		*value = data->rssi;
+		break;
+
+	case CELLULAR_SIGNAL_RSRP:
+	case CELLULAR_SIGNAL_RSRQ:
+		/* TODO: Validate and set values */
+		ret = -ENODATA;
+		break;
+	}
```
I had this in my PR, but since you're messing around here anyways. I think all switch-case should have default statements since `type` is equivalent to an int.

bjarki-trackunit replied:
Added defaults to both switches :)

bbilas commented on include/zephyr/drivers/cellular.h:
```diff
@@ -125,8 +180,21 @@ int cellular_get_supported_networks(const struct device *dev,
+	const struct cellular_driver_api *api = (const struct cellular_driver_api *)dev->api;
+
+	if (api->get_signal == NULL) {
+		return -ENOTSUP;
```
Shouldn't we use `ENOSYS` instead?

bjarki-trackunit replied:
Yes, updated to use ENOSYS

jukkar commented on include/zephyr/drivers/cellular.h:
```diff
@@ -96,8 +124,21 @@ enum cellular_modem_info_type {
  */
-int cellular_configure_networks(const struct device *dev, const struct cellular_network *networks,
-				uint8_t size);
+__syscall int cellular_configure_networks(const struct device *dev,
```
Does this API work properly when called from userspace? I have #65746 where I am trying to remove the syscall calls from phy interface as userspace cannot access the device.
So is this really tested to work from userspace application or did you added the syscall because many of the other APIs have it?

bjarki-trackunit replied:
I added it with the intention of adding handlers later. I have removed it for now :)

ldenefle reviewed (approved):
LGTM
